homeassistant/components/hunterdouglas_powerview/cover.py

```diff
@@ -136,21 +169,51 @@ def __init__(self, coordinator, device_info, room_name, shade, name):
         self._is_opening = False
         self._is_closing = False
         self._last_action_timestamp = 0
+        self._power_connection = 0
+        if isinstance(self._shade.raw_data[SHADE_BATTERY_KIND], int):
+            self._power_connection = self._shade.raw_data[SHADE_BATTERY_KIND]
         self._scheduled_transition_update = None
-        self._current_hd_cover_position = MIN_POSITION
+        self._current_hd_cover_primary = MIN_POSITION
+        self._current_hd_cover_secondary = MIN_POSITION
         if self._device_info[DEVICE_MODEL] != LEGACY_DEVICE_MODEL:
             self._attr_supported_features |= CoverEntityFeature.STOP
         self._forced_resync = None
 
     @property
     def extra_state_attributes(self):
         """Return the state attributes."""
-        return {STATE_ATTRIBUTE_ROOM_NAME: self._room_name}
+        return {
+            STATE_ATTRIBUTE_ROOM_NAME: self._room_name,
+            STATE_ATTRIBUTE_POWER_CONN: SHADE_POWER_CONN_LIST[self._power_connection],
+        }
+
+    @property
+    def get_position_primary(self):
+        """Access shade position directly from co-ordinator."""
+        if self.coordinator.data is None:
+            return self._current_hd_cover_primary
+        return self.coordinator.data[self._shade.id][ATTR_POSITION_DATA][ATTR_POSITION1]
```

Comment:
```python3

@property
def position_data(self):
    """Return the position data."""
    return self.coordinator.data[self._shade.id][ATTR_POSITION_DATA]
````

Reply:
Since we refresh each shade at startup the data isn't in the coordinator yet.  You could inject it with something like this:
```
diff --git a/homeassistant/components/hunterdouglas_powerview/cover.py b/homeassistant/components/hunterdouglas_powerview/cover.py
index b0fb9c6c73..085bc9c9e9 100644
--- a/homeassistant/components/hunterdouglas_powerview/cover.py
+++ b/homeassistant/components/hunterdouglas_powerview/cover.py
@@ -31,6 +31,7 @@ from homeassistant.config_entries import ConfigEntry
 from homeassistant.core import HomeAssistant, callback
 from homeassistant.helpers.entity_platform import AddEntitiesCallback
 from homeassistant.helpers.event import async_call_later
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
 
 from .const import (
     COORDINATOR,
@@ -81,7 +82,7 @@ async def async_setup_entry(
     room_data = pv_data[PV_ROOM_DATA]
     shade_data = pv_data[PV_SHADE_DATA]
     pv_request = pv_data[PV_API]
-    coordinator = pv_data[COORDINATOR]
+    coordinator: DataUpdateCoordinator = pv_data[COORDINATOR]
     device_info = pv_data[DEVICE_INFO]
 
     entities = []
@@ -100,6 +101,7 @@ async def async_setup_entry(
                 name_before_refresh,
             )
             continue
+        coordinator.data[shade.id] = shade.raw_data
         room_id = shade.raw_data.get(ROOM_ID_IN_SHADE)
         room_name = room_data.get(room_id, {}).get(ROOM_NAME_UNICODE, "")
         entities.extend(
```

Reply:
@bdraco setting it that way works - so are we ok with the catches for the likes of https://github.com/home-assistant/core/pull/62788#discussion_r868016850 when the coordinator is empty for the first 60 seconds ?

Reply:
We should take the guards out since we know the data will be there now with the above change.

Reply:
My bad - the ‘that works’ was referring to unique_id - I will need to test this one out and then assuming it works remove the guards